packages/vite/src/node/plugins/esbuild.ts

```diff
@@ -298,10 +298,19 @@ export function resolveEsbuildTranspileOptions(
   // pure annotations and break tree-shaking
   // https://github.com/vuejs/core/issues/2860#issuecomment-926882793
   const isEsLibBuild = config.build.lib && format === 'es'
+  const esbuildOptions = config.esbuild || {}
   const options: TransformOptions = {
-    ...config.esbuild,
+    ...esbuildOptions,
     target: target || undefined,
-    format: rollupToEsbuildFormatMap[format]
+    format: rollupToEsbuildFormatMap[format],
+    // the final build should always support dynamic import and import.meta.
+    // if they need to be polyfilled, plugin-legacy should be used.
+    // plugin-legacy detects these two features when checking for modern code.
+    supported: {
+      'dynamic-import': true,
+      'import-meta': true,
+      ...esbuildOptions.supported
+    }
```

Comment:
This part is the same as in the optimizer where the user can set `optimizeDeps.esbuildOptions.supported` to override too, even if it doesn't make sense. I think it might still be helpful if someone wants to override it for some reason. Though I don't have a hard opinion on this, if we want to change this to be stricter, we should change it for the optimizer too.